File: jupyter2/ps/variables/triangle_area/ml/triangle_area.ml
(** begin my answer *)

(** begin hidden *)
let triangle_area x1 y1 x2 y2 x3 y3 =
  let dx21 = x2 -. x1 in
  let dy21 = y2 -. y1 in
  let dx31 = x3 -. x1 in
  let dy31 = y3 -. y1 in
  0.5 *. abs_float (dx21 *. dy31 -. dx31 *. dy21);;
(** end hidden *)
(** end my answer *)

let main () =
  assert (abs_float (triangle_area 0.0 0.0 1.0 0.0 0.0 1.0 -. 0.5) < 1.0e-5);
  assert (abs_float (triangle_area 0.0 0.0 2.0 0.0 0.0 2.0 -. 2.0) < 1.0e-5);
  assert (abs_float (triangle_area 1.0 1.0 4.0 1.0 1.0 5.0 -. 6.0) < 1.0e-5);
  Printf.printf "OK\n"
;;

main()

File: jupyter2/ps/variables/triangle_area/go/triangle_area.go
package main
import "fmt"
import "math"

/** begin my answer */

/** begin hidden */
func triangleArea(x1, y1, x2, y2, x3, y3 float64) float64 {
	dx21 := x2 - x1
	dy21 := y2 - y1
	dx31 := x3 - x1
	dy31 := y3 - y1
	return 0.5 * math.Abs(dx21*dy31 - dx31*dy21)
}
/** end hidden */
/** end my answer */
func main() {
	if !(math.Abs(triangleArea(0,0,1,0,0,1) - 0.5) < 1.0e-5) { panic("wrong") }
	if !(math.Abs(triangleArea(0,0,2,0,0,2) - 2.0) < 1.0e-5) { panic("wrong") }
	if !(math.Abs(triangleArea(1,1,4,1,1,5) - 6.0) < 1.0e-5) { panic("wrong") }
	fmt.Println("OK")
}
